### PR TITLE
Add Skip() and Reset() methods to LittleEndian::Reader

### PR DESCRIPTION
**Pull request by holbrookt**
#### Problem
Reader has no methods for Skip() or Reset(), which may lead users to do potentially unsafe buffer size calculations, such as in #4748.

This is a follow-up to a comment in #4765.

 #### Summary of Changes
- add Skip() method to advance the Reader forward a number of octets
- add Reset() method to return the Reader to the start pointer it was initialized with

### CODE REVIEW

**Comment by andy31415** [comment]
Do we have a usage for this?

I am a bit weary for 'reset' because when getting such a buffer it  may not be clear to what it resets.
Skip seems to not be useful unless the buffer is already initialzed/it has already been written to.

I would like to understand use cases and if we could instead use existing functionality (e.g. one can allways use operator= to reset to a given buffer, but then it has to have knowledge of buffer and size).

**Comment by andy31415** [comment]
> Do we have a usage for this?
> 
> I am a bit weary for 'reset' because when getting such a buffer it may not be clear to what it resets.
> Skip seems to not be useful unless the buffer is already initialzed/it has already been written to.
> 
> I would like to understand use cases and if we could instead use existing functionality (e.g. one can allways use operator= to reset to a given buffer, but then it has to have knowledge of buffer and size).

Realized this is for reader ... so skip seems to make sense if data is not required. Maybe SkipBytes is a clearer name.
Still not convinced about reset.

**Comment by holbrookt** [comment]
> > Do we have a usage for this?
> > I am a bit weary for 'reset' because when getting such a buffer it may not be clear to what it resets.
> > Skip seems to not be useful unless the buffer is already initialzed/it has already been written to.
> > I would like to understand use cases and if we could instead use existing functionality (e.g. one can allways use operator= to reset to a given buffer, but then it has to have knowledge of buffer and size).
> 
> Realized this is for reader ... so skip seems to make sense if data is not required. Maybe SkipBytes is a clearer name.
> Still not convinced about reset.

The use case for this came from #4765 specifically because I wrote some code in `TestBdxTransferSession.cpp` that would inspect the contents of a PacketBuffer containing a StatusReport message. I needed to read the header and data in the PacketBuffer but not modify the Start pointer so that the message could then be passed to the "peer" node in the test.

In non-test code, this may not be a use case. I assume that it almost all cases, the header of a message will be consumed and then the data will be inspected and there will likely not be a need to reset the Start pointer and consume the header again. But it will make the test code safer and there may be other use cases for being able to reset/skip a Reader.

Would it be more clear to change `Reset()` to `Rewind()` or `BackToStart()`?

**Review by bzbarsky-apple — COMMENTED** [review]
Do we have a use case for Reset?  I'd avoid adding it, and the corresponding extra code, until we decide we really need it.

**Comment by bzbarsky-apple** [comment]
Fwiw, we should probably add the uses together with the methods; then it will be clearer why they are useful.

**Comment by holbrookt** [comment]
> Fwiw, we should probably add the uses together with the methods; then it will be clearer why they are useful.

Just pushed a commit that includes the use case for `Skip()`. There's currently no pending use case for `Reset()` that I know of in BDX. Maybe @tcarmelveilleux or @mspang know of a use since they suggested it?

**Review by bzbarsky-apple — COMMENTED** [review]
Right, the use case for Skip seems reasonable.  I'd rather we not add Reset, and its attendant extra member and whatnot until we have an actual use.

**Comment by kpschoedel** [comment]
> Maybe SkipBytes is a clearer name.

`Writer` already has `Skip`.

**Comment by kpschoedel** [comment]
`Writer` temporarily had `Reset()` for use by the `PacketBufferWriter` subclass, but I removed it again because copy assignment worked just as well, i.e. compilers are smart enough that `writer = Writer(…)` just resets the LHS.

**Comment by bzbarsky-apple** [comment]
> `Writer` already has `Skip`.

True, though it's unused (even in tests), so not sure why it's there...

**Comment by tcarmelveilleux** [comment]
> Do we have a use case for Reset? I'd avoid adding it, and the corresponding extra code, until we decide we really need it.

@bzbarsky-apple @andreilitvin The amount of extra code is minimal. The ability to rewind to start of a buffer to re-read it rather than create a new reader over the same buffer (which will involve more instructions) is a symmetry that has yet to be taken advantage of, but mostly because a lot of cases of manual buffer management still exist.

If you have a reader, a Reset() is a common method. It exists in our internal buffer reader APIs and is trivial.

Trevor wrote the tests and it's an inlined header method so if it's never used, it costs nothing.

**Comment by bzbarsky-apple** [comment]
> Trevor wrote the tests and it's an inlined header method so if it's never used, it costs nothing.

There's a nonzero cost per constructor invocation because of the extra state that needs to get stored, no?  That's why there's a codesize increase even though nothing is using the reset method....

If it were truly zero-cost-unless-used I would be a lot happier with it.

**Comment by bzbarsky-apple** [comment]
Maybe the right thing to do is to drop the new `mBufLen` member and have `Reset` do `mAvailable += OctetsRead();` before resetting mReadPtr?  I think that would make this actually zero-cost.

**Comment by andy31415** [comment]
> > Do we have a use case for Reset? I'd avoid adding it, and the corresponding extra code, until we decide we really need it.
> 
> @bzbarsky-apple @andreilitvin The amount of extra code is minimal. The ability to rewind to start of a buffer to re-read it rather than create a new reader over the same buffer (which will involve more instructions) is a symmetry that has yet to be taken advantage of, but mostly because a lot of cases of manual buffer management still exist.
> 
> If you have a reader, a Reset() is a common method. It exists in our internal buffer reader APIs and is trivial.
> 
> Trevor wrote the tests and it's an inlined header method so if it's never used, it costs nothing.

@tcarmelveilleux I would argue that reset in this context can cause more confusion. Consider this code:

```cpp

void ProcessData(Reader reader) {
   int data = reader.Read8();
   if (data == 0xFF) {  // if starting with 0xFF, data is 16-bit
      reader.Reset();
      data = reader.Read16();
   } 
   return data;
}
```

It "feels" ok (read something, need to re-read so reset and read again). But it is undefined behaviour because Reader buffer and length are not specified (function may read in the middle of a buffer).

I agree that Reset is very convenient if you create a buffer, do stuff and just need to go totally back. 

How about having an equivalent of Tell() and Seek() ?

so you can:

```cpp
auto pos = reader.Tell();

 if (data == 0xFF) {  // if starting with 0xFF, data is 16-bit
    reader.Seek(pos);
    data = reader.Read16();
 } 
 ```

this also allows for 0 cost in constructors. 
In the current implementation Tell/Seek could be simultated with operator=.

**Comment by tcarmelveilleux** [comment]
> > > Do we have a use case for Reset? I'd avoid adding it, and the corresponding extra code, until we decide we really need it.
> > 
> > 
> > @bzbarsky-apple @andreilitvin The amount of extra code is minimal. The ability to rewind to start of a buffer to re-read it rather than create a new reader over the same buffer (which will involve more instructions) is a symmetry that has yet to be taken advantage of, but mostly because a lot of cases of manual buffer management still exist.
> > If you have a reader, a Reset() is a common method. It exists in our internal buffer reader APIs and is trivial.
> > Trevor wrote the tests and it's an inlined header method so if it's never used, it costs nothing.
> 
> @tcarmelveilleux I would argue that reset in this context can cause more confusion. Consider this code:
> 
> ```c++
> int ProcessData(Reader reader) {
>    int data = reader.Read8();
>    if (data == 0xFF) {  // if starting with 0xFF, data is 16-bit
>       reader.Reset();
>       data = reader.Read16();
>    } 
>    return data;
> }
> ```
> 
> It "feels" ok (read something, need to re-read so reset and read again). But it is undefined behaviour because Reader buffer and length are not specified (function may read in the middle of a buffer).
> 
> I agree that Reset is very convenient if you create a buffer, do stuff and just need to go totally back.
> 
> How about having an equivalent of Tell() and Seek() ?
> 
> so you can:
> 
> ```c++
> auto pos = reader.Tell();
> int data = reader.Read8();
> if (data == 0xFF) {  // if starting with 0xFF, data is 16-bit
>    reader.Seek(pos);
>    data = reader.Read16();
> } 
> ```
> 
> this also allows for 0 cost in constructors.
> In the current implementation Tell/Seek could be simultated with operator=.

There is more conversation than code. It means we have to simplify. Let's keep `Skip()` and remove `Reset()` for now.

**Comment by holbrookt** [comment]
> There is more conversation than code. It means we have to simplify. Let's keep `Skip()` and remove `Reset()` for now.

Agreed. We can revisit this if a use case arises for Reset/Seek

File: src/lib/support/BufferReader.h [hunk 1]
```diff
@@ -156,6 +156,23 @@ class Reader
...
+    /**
+     * Advance the Reader forward by the specified number of octets.
+     *
+     * @param len The number of octets to skip.
+     *
+     * @note If the len argument is greater than the number of available octets
+     *       remaining, the Reader will advance to the end of the buffer
+     *       without entering a failed-status state.
+     */
+    Reader & Skip(uint16_t len)
```
Comment (tcarmelveilleux) [inline src/lib/support/BufferReader.h:168]: Replace uint16_t with size_t please

**Review by tcarmelveilleux — APPROVED** [review]

File: src/lib/support/BufferReader.h (hunk 1 above)
Comment (andy31415) [reply src/lib/support/BufferReader.h:168]: mAvailable is currently uint16_t.

if you want, we can define a size_type in the class, however I believe uint16_t is correct here.

**Review by andy31415 — APPROVED** [review]

File: src/lib/support/BufferReader.h (hunk 1 above)
Comment (holbrookt) [reply src/lib/support/BufferReader.h:168]: The signature of the constructor is `Reader(const uint8_t * buffer, uint16_t buf_len)` - does that `uint16_t` also need to be changed to `size_t`

**Review by bzbarsky-apple — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Review by kpschoedel — APPROVED** [review]

**Pull request merged**